src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs

```diff
@@ -367,6 +367,102 @@ private void SetInputPullModePi4(int pinNumber, PinMode mode)
+        /// </summary>
+        /// <param name="pinNumber">Pin to set</param>
+        /// <param name="altMode">Mode to set</param>
+        internal void SetAlternatePinMode(int pinNumber, int altMode)
+        {
+            Initialize();
+            /*
+             * There are 6 registers(4-byte ints) that control the mode for all pins. Each
+             * register controls the mode for 10 pins. Each pin uses 3 bits in the register
+             * containing the mode.
+             */
+
+            // Define the shift to get the right 3 bits in the register
+            int shift = (pinNumber % 10) * 3;
+            // Gets a pointer to the register that holds the mode for the pin
+            uint* registerPointer = &_registerViewPointer->GPFSEL[pinNumber / 10];
+            uint register = *registerPointer;
+            // Clear the 3 bits to 0 for the pin Number.
+            register &= ~(0b111U << shift);
+            // Set the 3 bits to the desired mode for that pin.
+            uint modeBits = 0; // Default: Gpio input
+
+            switch (altMode)
+            {
+                case 0:
+                    modeBits = 0b100;
+                    break;
+                case 1:
+                    modeBits = 0b101;
+                    break;
+                case 2:
+                    modeBits = 0b110;
+                    break;
+                case 3:
+                    modeBits = 0b111;
+                    break;
+                case 4:
+                    modeBits = 0b011;
+                    break;
+                case 5:
+                    modeBits = 0b010;
+                    break;
+            }
+
+            register |= (modeBits) << shift;
+            *registerPointer = register;
+        }
+
+        /// <summary>
+        /// Returns the currently set pin mode by directly reading the hardware
+        /// </summary>
+        /// <param name="pinNumber">Pin number</param>
+        /// <returns>(Alternate) Pin mode. 0 = Alt0, 1= Alt1... -1 Gpio Input, -2 Gpio Output</returns>
+        internal int GetAlternatePinMode(int pinNumber)
+        {
+            Initialize();
+            /*
+             * There are 6 registers(4-byte ints) that control the mode for all pins. Each
+             * register controls the mode for 10 pins. Each pin uses 3 bits in the register
+             * containing the mode.
+             */
+
+            // Define the shift to get the right 3 bits in the register
+            int shift = (pinNumber % 10) * 3;
+            // Gets a pointer to the register that holds the mode for the pin
+            uint* registerPointer = &_registerViewPointer->GPFSEL[pinNumber / 10];
+            uint register = *registerPointer;
+            // get the three bits of the register
+            register = (register >> shift) & 0b111;
+
+            switch (register)
+            {
+                case 0b000:
+                    // Input
+                    return -1;

```

Comment:
what are negative values, you didn't allow them for setter

Reply:
also, assuming we go with IGpioPin design this could be an implementation detail of RaspberryPi3Pin.PinMode

Reply:
Similar problem as with the enum above: I was intending to use 0-based addressing, so that Alt0 is 0, Alt1 is 1, etc. Then Input and Output fall out of that scheme, while technically they belong to the same set (There are 7 possible modes (3 bit): Input, Output, Alt0, Alt1, Alt2, Alt3, Alt4, Alt5 for each pin). 

A class type could work around that as well.